### PR TITLE
rm problematic variables under beta redexes for evar instantiation

### PR DESCRIPTION
**Pull request by Tragicus**
Extracted from https://github.com/coq/coq/pull/19822 as suggested [here](https://github.com/coq/coq/pull/19822#pullrequestreview-2433929513). 
The unification algorithm sometimes fails on problems of the form ?a = t because variables that are not in the scope of ?a appear in t. We can easily solve the problem when the variable appears in a beta-redex, as in (fun _ => t) x.

Fixes / closes #????

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [ ] Added / updated **documentation**.
  
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make doc_gram_rsts`.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by JasonGross** [comment]
https://github.com/coq/coq/blob/ac0fd362a8296285e6aeab394a9012741c1163b4/test-suite/bugs/bug_3209.v#L23-L74 needs to be fixed
@coqbot run full ci

**Comment by JasonGross** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by JasonGross** [comment]
CI seems to pass, let's see what the bench says

**Comment by JasonGross** [comment]
Bench results look like mostly noise, especially from the CPU column.  As far as I'm concerned, this could be good to go with just needs a changelog and a review from @coq/pretyper-maintainers.  @Tragicus , could you add a test-suite example that fails on master and passes with this change?

File: doc/changelog/02-specification-language/19833-evd-inst-beta.rst
Comment (JasonGross) [inline doc/changelog/02-specification-language/19833-evd-inst-beta.rst:2]: Be more declarative so users who don't know the algorithm can still understand the change
```suggestion
  Unification will now reduce beta-redexes when necessary to instantiate evars
```

**Comment by JasonGross** [comment]
cc @coq/pretyper-maintainers , does this look good?

**Comment by ppedrot** [comment]
Explicitly ping @mattam82

**Comment by JasonGross** [comment]
> I thought it was quite precise. Do you have a suggestion?

Hm, I guess the commit message disagrees with the changelog, and opens up a question: can unification solve a problem like `(fun f => f (fun a b => a)) (fun g => g x y) = ?e` when `y` is absent from `?e`?

If you want to improve the commit message, I'd suggest using the changelog message for the title of the commit (if it's accurate), and adding a sentence like "When unification fails to solve a problem of the form `(fun x => ...) ... =?= ?e`, we now try again with the weak-head beta-normal form of the term." (Is this accurate?  Does it extend to problems where we are not instantiating an evar but instead unifying two terms?  Is that case already handled?)

But I think generally commit messages on Coq are not held to a very high standard in general, so definitely feel free to leave the message as-is if you'd prefer.

**Comment by ppedrot** [comment]
@Tragicus this kind of changes to the unification algorithm can subtly break things. If the message is not much more explicit about what is going on in the change, there is a risk that 1. we forget the motivation when git blaming and 2. whoever is trying to bisect is going to be quite confused.

**Comment by ppedrot** [comment]
Basically a variant of what's written in the top post would make a good commit log.

**Comment by Tragicus** [comment]
Is that better?

**Comment by ppedrot** [comment]
The formatting is a bit weird. Normally git commit messages are made of a one-line title followed by a blank line and then the body of the message. I assume it's not very important though.

**Comment by Tragicus** [comment]
Ah, there was our misunderstanding, I am used to very short one-line commit messages. It does not cost anything to try to make it right, so there it is.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@mattam82 are you fine with the change?

**Comment by ppedrot** [comment]
@coqbot run full ci

**Review by mattam82 — APPROVED** [review]
Yes, I'm happy with the change (discussed with @Tragicus last week).

**Comment by ppedrot** [comment]
Then @coqbot merge now

**Pull request merged**
